src/wrapper-types.jl

```diff
@@ -57,58 +57,82 @@ function wraps_type end
+    else
+        kwarg
+    end
+end
+
+# helper function to extract argument names from expressions
+function expr_argname(arg)
+    if arg isa Expr && (arg.head == :(::) || arg.head == :(...))
+        arg.args[1]
+    elseif arg isa Expr
+        error("$arg not supported as an argument")
+    else
+        arg
+    end
+end
+
 function wrap_func_expr(mod, expr)
     @assert expr.head == :function || (expr.head == :(=) &&
                                        expr.args[1] isa Expr &&
                                        expr.args[1].head == :call)
 
     def = splitdef(expr)
-
-    sig = expr.args[1]
     body = def[:body]
-
     fname = def[:name]
     args = get(def, :args, [])
     kwargs = get(def, :kwargs, [])
+    args_names = expr_argname.(args)
+    kwargs_names = expr_kwargname.(kwargs)
+    
+    wrap_func_expr(mod, fname, args, kwargs, args_names, kwargs_names, Symbol[], body)
+end
 
-    impl_name = Symbol(fname,"_", hash(string(args)*string(kwargs)))
-
-    function kwargname(kwarg)
-        if kwarg isa Expr && kwarg.head == :kw
-            kwarg.args[1]
-        elseif kwarg isa Expr && kwarg.head == :(...)
-            kwarg.args[1]
-        else
-            kwarg
-        end
-    end
-
-    function argname(arg)
-        if arg isa Expr && (arg.head == :(::) || arg.head == :(...))
-            arg.args[1]
-        elseif arg isa Expr
-            error("$arg not supported as an argument")
-        else
-            arg
-        end
-    end
-
-    names = vcat(argname.(args), kwargname.(kwargs))
-
-    function type_options(arg)
+function wrap_func_expr(
+    mod, fname, args, kwargs, args_names, kwargs_names, whereparams, body;
+    abstract_arg_types=nothing
+)
+    names = vcat(args_names, kwargs_names)
+    
+    function type_options(wparams, arg, arg_ind)
+        pmod = parentmodule(mod)
+        atype = isnothing(abstract_arg_types) ? Any : abstract_arg_types[arg_ind]
         if arg isa Expr && arg.head == :(::)
-            T = Base.eval(mod, arg.args[2])
+            T = Base.eval(mod, quote
+                let $(Symbol(pmod)) = $(pmod); # make name of parent module available in eval scope   
+                    #=
+                    NOTE
+                    `typeintersect` is important here for consecutive calls to `specialize_methods` 
+                    with conceptually different super types. 
+                    E.g.: Consider we first specialize `*(::AbstractMatrix, ::AbstractVector)` to 
+                    redirect to `_matvec`, and then `*(::AbstractMatrix, ::AbstractMatrix)` to 
+                    redirect to `_matmul`. If we encounter some existing method for `*` which accepts
+                    an `AbstractMatrix` and an `VecOrMat` (type union), then we accidentally redirect 
+                    a matrix-vector-product to `_matmul` without `typeintersect`.
+                    =#
+                    typeintersect($(atype), $(arg.args[2]) where {$(wparams...)})
```

Comment:
What happens if methods are added by a third party package after `@warpped` has executed?

Reply:
Then `specialize_methods` would have to be called again. This sort of load-order dependence is also observed in [Hyperspecialize](https://github.com/willow-ahrens/Hyperspecialize.jl), and the only way around it seems the Cassette approach.
Actually, the code in this pull request is not too different from what Hyperspecialize can do, except that some of the Matrix or Vector types occurring in method definitions can be super special, so I had to either traverse the whole type tree and gather every possible subtype for Hyperspecialize's `@concretize` or inspect the method signatures (what is done now).

However, there are a few things we can do with respect to the points in [your other comment](https://github.com/JuliaSymbolics/Symbolics.jl/pull/811#issuecomment-1419267234):  

> All relevant types in Base/stdlib should be made to work with the wrapped types defined in this package

`specialize_methods` accepts a list of modules with argument `mods`. When calling it in `__init__()` we can restrict ourselves to `LinearAlgebra` and other relevant standard modules.

> Taking this further if a new package wants to use its array or number types with Symbolics, wherever necessary, it should return the specializations manually or using a different macro.

> What happens if methods are added by a third party package after @warpped has executed?

Whenever a user has custom array types defined in a module, or there is a package with custom array types, `specialize_methods` can be called with `mods` set accordingly to automatically “widen”/wrap the relevant definitions to have a generic fallback. This way, the developer does not have to know that `Base.:(*)(::AbstractMatrix, ::AbstractVector)` should redirect to `_matmul` most of the time.
If, for example, I get a `StaticMatrix` somewhere in my computations, then I cannot use it with symbolic arrays:
```
using Symbolics
using StaticArrays
A = @SMatrix(rand(2,2)) # imagine this is obtained somewhere else
@variables (x::Real)[1:2]
A * x # ERROR: MethodError: *(::SMatrix{2, 2, Float64, 4}, ::Symbolics.Arr{Num, 1}) is ambiguous.
```
But after
```Symbolics.specialize_methods(StaticArrays)```
it just works.

For some widely used packages (such as StaticArrays) we could even do this ourselves with `@require`. Conversely, a package author could opt to place `Symbolics.specialize_methods(@__MODULE__)` in their `__init__`, though I have to think about scopes and `export` statements a bit more.

Going forward, I can try to benchmark the impact of `specialize_methods` in `__init__` (for different values of `mods`) and maybe devise some tests.